tachyon/crypto/sumcheck/sumcheck_prover.h

```diff
@@ -0,0 +1,180 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVER_H_
+#define TACHYON_CRYPTO_SUMCHECK_SUMCHECK_PROVER_H_
+
+#include <utility>
+#include <vector>
+
+#include "tachyon/crypto/sumcheck/sumcheck_prover_msg.h"
+#include "tachyon/crypto/sumcheck/sumcheck_proving_key.h"
+#include "tachyon/crypto/sumcheck/sumcheck_verifier_msg.h"
+#include "tachyon/math/polynomials/multivariate/linear_combination.h"
+#include "tachyon/math/polynomials/univariate/univariate_evaluations.h"
+
+namespace tachyon::crypto {
+
+// The prover argues for the sum of the polynomial over {0,1}^|num_variables_|.
+//
+// The polynomial is represented by a list of terms each containing a
+// coefficient and a list of |MultilinearDenseEvaluations|. Refer to
+// "tachyon/math/polynomials/multivariate/linear_combination.h" for more info.
+template <typename MLE, size_t MaxDegree>
+class SumcheckProver {
+ public:
+  constexpr static size_t kParallelFactor = 16;
+
+  using F = typename MLE::Field;
+  using Point = typename MLE::Point;
+  using Term = math::LinearCombinationTerm<F>;
+
+  SumcheckProver() = default;
+  explicit SumcheckProver(const SumcheckProvingKey<MLE>& key)
+      : num_variables_(key.verifying_key.num_variables),
+        max_evaluations_(key.verifying_key.max_evaluations),
+        terms_(key.terms),
+        flattened_ml_evaluations_(key.flattened_ml_evaluations) {
+    // Constants are not accepted
+    CHECK_NE(num_variables_, size_t{0});
+    randomness_.reserve(num_variables_);
+  }
+  explicit SumcheckProver(SumcheckProvingKey<MLE>&& key)
+      : num_variables_(key.verifying_key.num_variables),
+        max_evaluations_(key.verifying_key.max_evaluations),
+        terms_(std::move(key.terms)),
+        flattened_ml_evaluations_(std::move(key.flattened_ml_evaluations)) {
+    // Constants are not accepted
```

Comment:
It's not mat critical, though.

```suggestion
    // Constants are not accepted.
```